File: Leetcode/Python/_21.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list2 is None:
            return list1
        elif list1 is None:
            return list2

        if list1.val > list2.val:
            head = list2
            list2 = list2.next
        else:
            head = list1
            list1 = list1.next

        head.next = self.mergeTwoLists(list1, list2)

        return head
    
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        result = head 
        
        if list1 is None:
            return list2
        elif list2 is None:
            return list1
        elif list1 is None and list2 is None:
            return []
        
        while list1 and list2:
            if list1.val > list2.val:
                head.next = ListNode(list2.val)
                list2 = list2.next
                head = head.next 
            else:
                head.next = ListNode(list1.val)
                list1 = list1.next
                head = head.next 
                
        while list2:
            head.next = ListNode(list2.val)
            list2 = list2.next
            head = head.next 
        while list1:
            head.next = ListNode(list1.val)
            list1 = list1.next
            head = head.next 
            
        return result.next
    
class Solution:
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if l1 is None and l2 is None:
            return None
        elif l1 is None:
            return l2
        elif l2 is None:
            return l1
        else:
            listnode = ListNode()
            cursor = listnode
            while l1 and l2:
                if l1.val > l2.val:
                    cursor.next = ListNode(l2.val)
                    cursor = cursor.next
                    l2 = l2.next
                else:
                    cursor.next = ListNode(l1.val)
                    cursor = cursor.next
                    l1 = l1.next
            while l1:
                cursor.next = ListNode(l1.val)
                l1 = l1.next
                cursor = cursor.next
            while l2:
                cursor.next = ListNode(l2.val)
                l2 = l2.next
                cursor = cursor.next
        return listnode.next

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode()
        temp = head
        while list1 and list2:
            if list1.val < list2.val:
                temp.next = list1
                list1 = list1.next
            else:
                temp.next = list2
                list2 = list2.next
            temp = temp.next
        if list1:
            temp.next = list1
        if list2:
            temp.next = list2
        return head.next

File: Leetcode/cpp/21.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* dfs(ListNode* list1, ListNode* list2) {
        if (list1 != nullptr && list2 != nullptr) {
            ListNode * node;
            if (list1->val < list2->val) {
                node = new ListNode(list1->val);
                list1 = list1->next;
            } else {
                node = new ListNode(list2->val);
                list2 = list2->next;
            }
            node->next = dfs(list1, list2);
            return node;
        } else if ((list1 == nullptr && list2 != nullptr) || (list1 != nullptr && list2 == nullptr)) {
            ListNode * node;
            if (list1 == nullptr) {
                node = new ListNode(list2->val);
                node->next = dfs(list1, list2->next);
                return node;
            } else {
                node = new ListNode(list1->val);
                node->next = dfs(list1->next, list2);
                return node;
            }
        }
        return nullptr;
    }

    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr && list2 == nullptr) {
            return nullptr;
        } else if (list2 == nullptr && list1 != nullptr) {
            return list1;
        } else if (list1 == nullptr && list2 != nullptr) {
            return list2;
        }
        ListNode * ptr;
        if (list1->val < list2->val) {
            ptr = list1;
            list1 = list1->next;
        } else {
            ptr = list2;
            list2 = list2->next;
        }
        ptr->next = mergeTwoLists(list1, list2);
        return ptr;
        // return dfs(list1, list2);
    }
};
